test/unit/math/mix/functor/reduce_sum_test.cpp

```diff
@@ -0,0 +1,291 @@
+#include <stan/math/prim/meta.hpp>
+#include <test/unit/math/test_ad.hpp>
+
+#include <limits>
+#include <vector>
+
+std::ostream* msgs = nullptr;
+
+template <typename T, stan::require_stan_scalar_t<T>* = nullptr>
+T sum_(T arg) {
+  return arg;
+}
+
+template <typename EigMat, stan::require_eigen_t<EigMat>* = nullptr>
+auto sum_(EigMat&& arg) {
+  return stan::math::sum(arg);
+}
+
+template <typename Vec, stan::require_std_vector_t<Vec>* = nullptr>
+auto sum_(Vec&& arg) {
+  stan::scalar_type_t<Vec> sum = 0;
+  for (size_t i = 0; i < arg.size(); ++i) {
+    sum += sum_(arg[i]);
+  }
+  return sum;
+}
+
+struct sum_lpdf {
+  template <typename T, typename... Args>
+  inline auto operator()(std::size_t start, std::size_t end, T&& sub_slice,
+                         std::ostream* msgs, Args&&... args) const {
+    using return_type = stan::return_type_t<T, Args...>;
+
+    return sum_(sub_slice)
+           + sub_slice.size()
+                 * stan::math::sum(std::vector<return_type>{
+                       return_type(sum_(std::forward<Args>(args)))...});
+  }
+};
```

Comment:
We should put all these in a namespace instead of floating around

Reply:
But this is just test code... so it's fine to leave as is, no?

Reply:
@SteveBronder you asked about
> What do you mean pure values here?

which was in the context how to best do the deep copies.

What I had in mind was that it would likely be better to first convert all `var`s into their respective value and *then* use the converted value only data structure to cast everything what needs to be a `var` into a var. The thing is that we do many deep copies of `var`s. Doing this off plain value representations is probably faster as compared to making these copies off the `var`s themselves. The reason is that the value only representation is more compact in memory.

However, given where we are I would suggest to leave things as they are right now. The type logic is very nice right now and the above suggestion would require to overhaul that and if my intuition is here is right is not fully ensured. So that would need benchmarks and all that which is time better spend now on polishing this up and include in develop.